opm/simulators/linalg/ISTLSolverEbosFlexible.hpp

```diff
@@ -0,0 +1,214 @@
+/// number of cell variables.
+///
+/// The solvers and preconditioners used are run-time configurable.
+template <class TypeTag>
+class ISTLSolverEbosFlexible
+{
+    using SparseMatrixAdapter = typename GET_PROP_TYPE(TypeTag, SparseMatrixAdapter);
+    using VectorType = typename GET_PROP_TYPE(TypeTag, GlobalEqVector);
+    using Simulator = typename GET_PROP_TYPE(TypeTag, Simulator);
+    using Scalar = typename GET_PROP_TYPE(TypeTag, Scalar);
+    using MatrixType = typename SparseMatrixAdapter::IstlMatrix;
+#if HAVE_MPI
+    using Communication = Dune::OwnerOverlapCopyCommunication<int, int>;
+#endif
+    using SolverType = Dune::FlexibleSolver<MatrixType, VectorType>;
+
+
+public:
+    static void registerParameters()
+    {
+        FlowLinearSolverParameters::registerParameters<TypeTag>();
+    }
+
+    explicit ISTLSolverEbosFlexible(const Simulator& simulator)
+        : simulator_(simulator)
+        , comm_(nullptr)
+    {
+        parameters_.template init<TypeTag>();
+        prm_ = setupPropertyTree(parameters_);
+        extractParallelGridInformationToISTL(simulator_.vanguard().grid(), parallelInformation_);
+        detail::findOverlapRowsAndColumns(simulator_.vanguard().grid(), overlapRowAndColumns_);
+#if HAVE_MPI
+        if (parallelInformation_.type() == typeid(ParallelISTLInformation)) {
+            // Parallel case.
+            const ParallelISTLInformation* parinfo = boost::any_cast<ParallelISTLInformation>(&parallelInformation_);
+            assert(parinfo);
+            comm_.reset(new Communication(parinfo->communicator()));
+        }
+#endif
+    }
+
+    void eraseMatrix()
+    {
+    }
+
+    void prepare(SparseMatrixAdapter& mat, VectorType& b)
+    {
+#if HAVE_MPI
+        static bool firstcall = true;
+        if (firstcall && parallelInformation_.type() == typeid(ParallelISTLInformation)) {
+            // Parallel case.
+            const ParallelISTLInformation* parinfo = boost::any_cast<ParallelISTLInformation>(&parallelInformation_);
+            assert(parinfo);
+            const size_t size = mat.istlMatrix().N();
+            parinfo->copyValuesTo(comm_->indexSet(), comm_->remoteIndices(), size, 1);
+            firstcall = false;
+        }
+        makeOverlapRowsInvalid(mat.istlMatrix());
+#endif
+        // Decide if we should recreate the solver or just do
+        // a minimal preconditioner update.
+        const int newton_iteration = this->simulator_.model().newtonMethod().numIterations();
+        bool recreate_solver = false;
+        if (this->parameters_.cpr_reuse_setup_ == 0) {
+            recreate_solver = true;
+        } else if (this->parameters_.cpr_reuse_setup_ == 1) {
+            if (newton_iteration == 0) {
+                recreate_solver = true;
+            }
+        } else if (this->parameters_.cpr_reuse_setup_ == 2) {
+            if (this->iterations() > 10) {
+                recreate_solver = true;
```

Comment:
I will add comments. This particular version recreates the solver if the previous solve used more than 10 iterations. It may therefore be recreated anywhere from every newton iteration or never, depending on iteration counts.